<H1 class="title" align="center">Frequently Asked Questions</H1>

<br>
<p align="center">
This is an updated copy of the complete Robetta documentation. Services that are not provided on this website are linked to the main <A href="http://robetta.bakerlab.org">Robetta</A> website.
</p>
<br>
<P>
<table border=0 cellpadding=3 cellspacing=0>
<tr bgcolor="#eeeeee"><td><LI> <a href="#whatis_robetta"><b>What is Robetta?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#whatis_fraglib"><b>What are Fragment Libraries?</b></a></td></tr>

<tr bgcolor="#eeeeee"><td><LI> <a href="#whatis_alascan"><b>What is Interface Alanine Scanning?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#whatis_ginzu"><b>What is Ginzu?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#submit"><b>How do I submit a sequence to the structure prediction server?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#fragsubmit"><b>How do I submit a sequence to the fragment server?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#alascansubmit"><b>How do I submit a sequence to the interface alanine scanning server?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#removingjobs"><b>How do I remove my job from the structure prediction server?</b></a></td></tr>

<tr bgcolor="#eeeeee"><td><LI> <a href="#mirror"><b>What's a mirror?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#process_time"><b>How long does a job take to process?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#one_job"><b>Why am I only allowed one job at a time?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#data"><b>How much time do I have before my job is removed and my data is lost?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#rosetta_license"><b>Can I run Rosetta myself?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#commercial_use"><b>Why can't commercial users submit jobs to Robetta?</b></a></td></tr>

<tr bgcolor="#eeeeee"><td><LI> <a href="#public_results"><b>Why are all my results visible to everyone?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#rasmol"><b>How do I configure my browser to view structures with RasMol?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#model_order"><b>Why do I get more than one model?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#domain_confidence"><b>What does the domain confidence mean?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#ksync_detail"><b>How do I interpret the K*Sync detailed alignment?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#long_sequences"><b>Are there any issues with long sequences?</b></a></td></tr>

<tr bgcolor="#eeeeee"><td><LI> <a href="#short_sequences"><b>Are there any issues with short sequences?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#ai_semantics"><b>What is the difference between Ab Initio and De Novo Modeling?</b></a></td></tr>
<tr bgcolor="#eeeeee"><td><LI> <a href="#cm_semantics"><b>What is the difference between Comparative and Homology Modeling?</b></a></td></tr>
</table>
</P>

<BR><BR>


<a name="whatis_robetta">
<H1 class="title">Q. What is Robetta? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>

<UL>
<P>
Robetta is a full-chain protein structure prediction server.  It parses protein chains into putative domains with the Ginzu protocol, and models those domains either by homology modeling or by ab initio modeling.  Other services of the server, in addition to Domain Parsing and 3-D Modeling, include Fragment Library generation and Interface Alanine Scanning.
</UL>
</P>

<br>
<a name="whatis_fraglib">
<H1 class="title">Q. What are Fragment Libraries? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
Fragment Libraries are the pieces of experimentally determined structures that Rosetta uses to guide the search of conformational space when predicting structures using the ab initio protocol, as well as longer loop conformations in homology models.
</UL>
</P>

<br>
<a name="whatis_alascan">
<H1 class="title">Q. What is Interface Alanine Scanning? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
  Interface Alanine Scanning attempts to estimate the energetic contribution
  to the binding free energy provided by each residue at a protein-protein
  interface. Briefly, interface alanine scanning uses a simple physical
  model to score a series of protein-protein interfaces in which contact
  residues are individually replaced with alanine. After each computational
  alanine mutation, the resulting binding energy is calculated.
  <p>
  The input consists of a three-dimensional structure of a protein-protein
  complex; output is a list of "hot spots," or amino acid side chains that
  are predicted to significantly destabilize the interface when mutated to
  alanine, analogous to the results of experimental alanine-scanning
  mutagenesis. 79% of hot spots and 68% of neutral residues were correctly
  predicted in a test of 233 mutations in 19 protein-protein complexes. A
  single interface can be analyzed in minutes. The computational methodology
  has been validated by the successful design of protein interfaces with new
  specificity and activity and has yielded new insights into the mechanisms
  of receptor specificity and promiscuity in biological systems.
</UL>
</P>

<br>
<a name="whatis_ginzu">
<H1 class="title">Q. What is Ginzu? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
Ginzu is a protocol that attempts to determine the regions of a protein chain that will fold into globular units, called "domains".  It scans the protein chain sequence with successively less confident methods of detection to determine any homologs with experimentally determined structures, starting with PDB-BLAST, and followed by the more remote fold-detection method HHSEARCH.  After any homologs are identified, a search of remaining regions is done with HMMER against the Pfam-A protein family database.  Lastly, the PSI-BLAST multiple sequence alignment is used to assign regions of increased likelihood of possessing a contiguous domain based on sequence clusters.  The final step consists of selecting cut-points between the domains (and possibly defining new domains based on the strongest cutpoints for any remaining long stretches of the sequence that have not already matched a homolog with a structure or Pfam-A) using the PSI-BLAST MSA.
</UL>
</P>

<br>
<a name="submit">
<H1 class="title">Q. How do I submit a sequence to the structure prediction server? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>

<UL>
<P>
To submit a sequence to the structure server, you must be a registered user. To register, <a href="http://robetta.bakerlab.org/register.jsp">click here</a>. 
If you are a registered user, go to the <a href="http://robetta.bakerlab.org/submit.jsp">Structure Server</a> submission form and do the following:
<OL>
<LI type="1"> Select a prediction type. 
<LI type="1"> Enter your username or registered email address.
<LI type="1"> Enter a target name for your sequence.
<LI type="1"> Paste your <a href="alascan2.py?query=data_formats#fasta">fasta</a> sequence in the text area or upload your <a href="alascan2.py?query=data_formatsts#fasta">fasta</a> file.

<LI type="1"> Fill out optional fields if desired.
<LI type="1"> Click the Submit button.
</OL>
</UL>
</P>

<br>
<a name="fragsubmit">
<H1 class="title">Q. How do I submit a sequence to the fragment server?  &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
To submit a sequence to the fragment server, you must be a registered user. To register, <a href="http://robetta.bakerlab.org/register.jsp">click here</a>.
If you are a registered user, go to the <a href="http://robetta.bakerlab.org/fragmentsubmit.jsp">Fragment Server</a> submission form and do the following:

<OL>
<LI type="1"> Enter your username or registered email address.
<LI type="1"> Enter a target name for your sequence.
<LI type="1"> Paste your <a href="alascan2.py?query=data_formats#fasta">fasta</a> sequence in the text area or upload your <a href="alascan2.py?query=data_formats#fasta">fasta</a> file.
<LI type="1"> Complete optional fields if desired. If you are uploading constraints data, be sure the <a href="alascan2.py?query=data_formats">formats</a> 

are correct to prevent errors and wasted processor time.
<LI type="1"> Click the Submit button.
</OL>
</UL>
</P>


<br>
<a name="alascansubmit">
<H1 class="title">Q. How do I submit a sequence to the interface alanine scanning server?  &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
To submit a sequence to the interface alanine scanning server, you must be a registered user. To register, <a href="alascan2.py?query=register">click here</a>.
If you are a registered user, login to the <a href="alascan2.py?query=login">Alanine Scanning Server</a> and go to the submission form.

<OL>
<LI type="1"> Enter a job name for your sequence.
<LI type="1"> Upload your <a href="alascan2.py?query=data_formats#pdbcomplex">protein complex</a> file (must be PDB format).
<LI type="1"> Define the interface for alanine scanning by entering the chain ID's (as in the complex; case sensitive) involved in the interface and the interface partners to which they belong.
<LI type="1"> Optional: If you want results for specific interface side-chains, you can upload a <a href="alascan2.py?query=data_formats#mutations">Mutations List</a>. <a href="alascan2.py?query=data_formats#mutations">Click here</a> for file format. If there is a format error, all interface side-chains will be considered.

<LI type="1"> Click the Submit button.
</OL>
</P>
</UL>

<br>
<a name="removingjobs">
<H1 class="title">Q. How do I remove my job from the structure prediction server?  &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
To remove one or more of your jobs from the structure prediction server, follow these instructions:
<OL>
<LI type="1"> <a href="http://robetta.bakerlab.org/login.jsp">Login</a> with your username and password.

<LI type="1"> Go to the structure prediction <a href="http://robetta.bakerlab.org/queue.jsp">queue table</a>.
<LI type="1"> Select one or more checkboxes (far left column under "x") for jobs that you own and would like to remove.
<LI type="1"> Click the "Update Job(s)" button on the bottom of the page.
</OL>
</P>
<P>
<b><font color="red">All data will be lost, so be sure you retrieve your data before doing this.</font></b>
</P>
<P>
<b><font color="red">Jobs may be removed one week after they complete to conserve disk space.</font></b>

</P>
</UL>



<br>
<a name="mirror">
<H1 class="title">Q. What is a mirror?  &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
A Robetta mirror is a cluster of computers that have been hooked into the Robetta system to perform the actual work of processing the targets.  There are currently two mirrors, generously provided by Charlie Strauss of the Los Alamos National Laboratory and Richard Bonneau of the Institute for Systems Biology.
</P>
</UL>

<br>
<a name="process_time">
<H1 class="title">Q. How long does a job take to process? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
Both Rosetta ab initio and comparative modeling must generate ensembles of decoys before selecting models from amongst the ensemble.  This process takes a considerable amount of computational resources, and therefore one can expect that, once your job has finished the Ginzu step and entered the ensemble generation, each domain will take a few hours (e.g. a 150 residue domain will take about 4 hours).  Other long waits of up to 12 hours may occur during the initial Ginzu step.
</P>
</UL>


<br>
<a name="one_job">
<H1 class="title">Q. Why am I only allowed one job at a time? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>

<UL>
<P>
Regrettably, due to the length of time it takes (on average) for jobs to complete, public users of the Robetta server are limited to one job at a time.  We hope that this will prevent exceptionally long queues from occuring.
</P>
</UL>


<br>
<a name="data">
<H1 class="title">Q. How much time do I have before my job is removed and my data is lost? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
We may remove domain and stucture prediction jobs and all their data one week after the date of completion to free disk space. 
This will be done at our own descretion.
</P>

</UL>


<br>
<a name="rosetta_license">
<H1 class="title">Q. Can I run Rosetta myself? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
Due to hardware resource limitations, we suggest that if you wish to run a large number of predictions and have ample hardware of your own, that you obtain the Rosetta suite of programs themselves (which are freely available to academic users by license, and may be commercially licensed as well).
</P>

<P>
  For more information <a href="http://depts.washington.edu/ventures/UW_Technology/Express_Licenses/Rosetta/">click here</a>.

</P>

</UL>


<br>
<a name="commercial_use">
<H1 class="title">Q. Why can't commercial users submit jobs to Robetta? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
In an effort to minimize overuse, we only permit non-commercial use of the Robetta server.  Additionally, Robetta uses methods can only be provided free-of-charge to academic users.  If you are a commercial user, and would like to use Rosetta, we suggest you pursue licensing Rosetta for use on your own hardware (see above).
</P>
</UL>

<br>
<a name="public_results">
<H1 class="title">Q. Why are all my results visible to everyone? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
As a public resource, we feel that the free-flow of information is important to maintain.  Additionally, <font color=red><b>we cannot take responsibility for providing the security necessary to protect users' results</b></font>.  Lastly, there may be more than one person who is interested in a particular target, and we want to avoid duplicate jobs.
</P>
</UL>


<br>

<a name="rasmol">
<H1 class="title">Q. How do I configure my browser to view structures with RasMol? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
Rasmol is a popular molecular graphics software package. If you do not have Rasmol, <A HREF="http://www.umass.edu/microbio/rasmol/getras.htm" target="_blank">click here</A>.
</P>
<P>
In theory, you should be able to configure a web browser to use Rasmol as a helper application to view PDB structures through the internet by 
setting the appropriate MIME type.
However, it is difficult, if not impossible, to set up Internet Explorer to do this. Therefore, we recommend using a Netscape browser and the following instructions.    
</P>
<P>
To configure your browser to view PDB structures &nbsp; <img src="../images/pdb.jpg" align=absmiddle border=0> &nbsp;, do the following:

</P>
<OL>
<LI type="1"> Go to your browser's Applications Preferences.
<LI type="1"> Add a new file type configured as:
    <UL>
        <LI type="disc"> <B>Description</B>: PDB files
        <LI type="disc"> <B>MIME Type</B>: chemical/x-pdb
        <LI type="disc"> <B>Suffixes</B>: .pdb,.ent
        <LI type="disc"> <B>Application</B>: xterm -e rasmol -pdb %s
    </UL>

</OL>
<P>
To configure your browser to view Rasmol scripts &nbsp; <img src="../images/rasmol.jpg" align=absmiddle border=0> &nbsp;, do the following:
</P>

<OL>
<LI type="1"> Go to your browser's Applications Preferences.
<LI type="1"> Add a new file type configured as:
    <UL>
        <LI type="disc"> <B>Description</B>: RasMol scripts
        <LI type="disc"> <B>MIME Type</B>: application/x-rasmol
        <LI type="disc"> <B>Suffixes</B>: (leave blank)
        <LI type="disc"> <B>Application</B>: xterm -e rasmol -script %s
    </UL>

</OL>

<P>
In the examples above, we use 'xterm -e rasmol' as the helper application command that launches Rasmol in an xterm terminal window (Unix/Linux platforms). 
You may substitute this with 'rasmac' or 'raswin.exe' depending on what platform you are using.
</P>

</UL>

<br>
<a name="model_order">
<H1 class="title">Q. Why do I get more than one model?  &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
There are several models that the Robetta server produces when you submit a sequence.  If Robetta determines that there is more than one domain (or if an ab initio portion of the sequence is too big to be modeled by the ab initio protocol), Robetta breaks up the query into putative domains and models each of them separately.  After doing so, it assembles the models into one contiguous chain.  This means you can examine your models either as a complete chain, or by clicking on the domain number in the Ginzu domain info box, you can examine the results for each individual domain.

<p>
Within each domain, there are several models.  In the case of ab initio predictions, the models are the cluster centers of the most populated clusters, with the exception of the last model, which is the lowest energy decoy that was not a member of the previously represented clusters.  In the case of homology modeling predictions, the second model is the model produced by the default K*Sync alignment, with the first, third, fourth, and fifth selected from the decoy ensemble by various energy discrimination methods.  In the case of twilight-zone reliability parent detections, the first 5 models are homology modeled, and models six-ten modeled using the de novo protocol.
</UL>
</P>


<br>
<a name="domain_confidence">
<H1 class="title">Q. What does the domain prediction confidence mean? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
The domain prediction confidence has different meanings depending on the method used to detect the region.  Those regions with a detected parent PDB structure are meant to have a confidence function that follows a similar trend.  The confidence value is derived based on the detection method in the following way:
<P>
  <table border=1 cellpadding=5 cellspacing=0>

    <tr><td>PDB-BLAST</td><td>homology</td><td>conf = -log(e-val)</td><td><nobr>e.g. e=.001  -> conf=3.0 (strong detection threshold)</td></tr>
        <tr><td>HHSEARCH</td><td>homology</td><td>conf = hhsearch_prob/42.5</td><td><nobr>e.g. prob=85.0 -> conf=2.0 (strong detection threshold)</td></tr>
<!--    <tr><td>FFAS03</td><td>homology</td><td>conf = -ffas_conf/10.0</td><td><nobr>e.g. f=-20.0 -> conf=2.0 (strong detection threshold)</td></tr>
    <tr><td>3D-Jury</td><td>homology</td><td>conf = jury_conf/25.0</td><td><nobr>e.g. j=50.0  -> conf=2.0 (strong detection threshold)</td></tr> -->
    <tr><td>Pfam</td><td>de novo</td><td>conf = -log(e-val)</td><td><nobr>e.g. e=.001  -> conf=3.0 (strong detection threshold)</td></tr>

    <tr><td>msa</td><td>de novo</td><td><nobr>conf = block_depth +</nobr><br><nobr>&nbsp;&nbsp;&nbsp; .001*block_occ +</nobr><br><nobr>&nbsp;&nbsp;&nbsp; .000001*e_val_pref +</nobr><br><nobr>&nbsp;&nbsp;&nbsp; .000000001*block_len</nobr></td><td>note: dominated by nr50 block depth</td></tr>
    <tr><td>cutpref</td><td>de novo</td><td>conf = 0</td><td>note: domain boundaries solely determined by sequence transitions, strongly predicted loop, occupancy, and distance from nearest block or terminus </td></tr>

  </table>
<P>
<table border=0 cellpadding=0 cellspacing=0><td>
The general trend with the homology modeling detections allows one to discriminate likely correct parents from improbable ones.  If the confidence for a parent PDB is >= 3.0, then it's almost certainly the right fold, and the model itself probably does a good job capturing the features of the structure.  Between 2.0 and 3.0, it's usually the right fold, but the model quality is likely to be reduced.  Between 1.0 and 2.0, the fold is still right more than half the time, but even so the models produced are often not as good as they could be in cases where the fold is correct, due to the difficulty that homology modeling faces at greater distance.  Therefore, in this extreme twilight-zone regime, Robetta also provides de novo models for such domains.
</td></table>
</UL>
</P>


<br>
<a name="ksync_detail">
<H1 class="title">Q. How do I interpret the K*Sync detailed alignment? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>

<P>
The K*Sync detailed color alignment view indicates more than just which
residues of the query align with which residues of the parent.  The structural
information that goes into the K*Sync alignments is also shown.  Information
for the query is shown on top, and that for the parent on the bottom.  Aligned
stretches are shown by blue bars between the query and the parent sequence.
Identical residues are optionally shown in black, or alternatively given the
color for aligned residues that are similar.  The color scheme for residue
classes is:
<P>
<table border=0 cellpadding=5 cellspacing=5>
<tr><td align=center bgcolor=#ffcc00>light orange</td><td>hydrophobic</td><td>(A,V,L,I,M,F,W,P)</td></tr>
<tr><td align=center bgcolor=#ff9900>dark orange</td><td>ring containing</td><td>(F,Y,W,H)</td></tr>
<tr><td align=center bgcolor=#ffccff>pink</td><td>hydrophilic</td><td>(K,R,D,E,H,Q,N,S,T)</td></tr>
<tr><td align=center bgcolor=#0000ff><font color=white>blue</font></td><td>basic</td><td>(K,R)</td></tr>

<tr><td align=center bgcolor=#ff0000><font color=white>red</font></td><td>acidic</td><td>(D,E)</td></tr>
<tr><td align=center bgcolor=#00ffff>cyan</td><td>turn and small</td><td>(P,G,A,S)</td></tr>
<tr><td align=center bgcolor=#ffff00>yellow</td><td>sulfur and small oxygen</td><td>(C,M,S,T)</td></tr>
</table>

<P>
Additionally shown are the structural information terms used by K*Sync.  They
include:

<P>

<UL>
<LI><b>ss_pred</b> and <b>ss_conf</b>: the 3-class secondary structure
predicted by the <a href="alascan2.py?query=doc#psipred">PSI-PRED</a> program.
The confidence of the prediction for that position is indicated by the height
of the <font color=#00ff00>green</font> bar.  The secondary structure classes
are helix=H, strand/sheet/extended=E, coil/loop=L.
<LI><b>ss_dssp</b>: is the DSSP secondary structure read from the parent
structure by <a href="alascan2.py?query=doc#dssp">DSSP</a> and collapsed to
3-classes (where {H,G} -> H and {E,B} -> E, and everything else is L).
Aligned positions that share the same secondary structure classification are
colored with <font color=#ff00ff>magenta</font> for H, 

<font color=#ffff00>yellow</font> for E, and <font color=#00ffff>cyan</font>
for L.
<LI><b>obl_msa</b> and <b>obl_str</b>: highly occupied positions in a
multiple sequence or multiple structural alignment are more likely obligate
to the fold.  The height of the orange bar in the <b>obl_msa</b> indicates
the degree of this occupancy for the query and the parent
<a href="alascan2.py?query=doc#psiblast">PSI-BLAST</a> multiple sequence
alignments, and the <b>obl_str</b> the occupancy of the parent in
a <a href="alascan2.py?query=doc#stradstack">StrAD-Stack</a> multiple
structural alignment of the parent structure with other non-redundant
experimental structures possessing the same fold.

</UL> 


</UL>
</P>


<br>
<a name="long_sequences">
<H1 class="title">Q. Are there any issues with long sequences? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
The Rosetta folding program itself is a FORTRAN program that has to be careful with its memory usage.  Extremely long sequences cannot be fit into memory, so any domain level models are limited to no more than 250 residues for the de novo protocol, and 600 residues for the comparative modeling protocol.
<p>
The de novo protocol suffers, as do all such methods, from a limitation in the ability to sample conformations available to the protein.  Larger targets that are high contact-order are more difficult to sample in a reasonable amount of computer time, and probably require much larger decoy ensembles than the Robetta server can afford to generate.  Therefore, we impose a de novo domain size limit of about 200 residues, which is clearly often incorrect, but necessary.  It is hoped that in such cases, features of the target are still captured by the models.
<p>

There is additionally a limit on the length of the full chain of about 1000 residues, so that the independently modeled domains may be assembled into a contiguous chain.
<!--<p>
Problems may also arise in cases where there are no homologous PDBs found in very long sequences, as the <a href="http://bioinfo.pl/">BioInfo</a> Meta server (where 3D-Jury is run) cannot accept extremely long queries.  Exercise caution!
-->
</UL>
</P>


<br>
<a name="short_sequences">
<H1 class="title">Q. Are there any issues with short sequences? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
The Fragment Library used by the Rosetta folding program is generated by using residue substitution profiles with PSI-BLAST to find similar fragment profiles.  Very short sequences (less than about 40 residues) are difficult to detect sequence homologs for with reliable confidence with PSI-BLAST, and therefore the fragments library may suffer.
<p>
Additionally, the Rosetta de novo protocol folds two divergent homologous sequences in addition to the query sequence.  If PSI-BLAST cannot find any sequence homologs with confidence due to a query being exceptionally short, Rosetta cannot take advantage of multiple homologs, and the modeling may suffer.

<p>
Perhaps most importantly in de novo modeling, the assumption is made that the target protein forms a soluble domain with a hydrophobic core.  Short sequences often do not fold up in this fashion, so the energy function used in Rosetta may incorrectly bias the structures to be more compact than they should be for short targets.
</UL>
</P>


<br>
<a name="ai_semantics">
<H1 class="title">Q. What is the difference between Ab Initio and De Novo Modeling? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
Ab initio structure prediction classically refers to structure prediction using nothing more than first-principles (i.e. physics).  De Novo is a more general term that refers to the greater category of methods that do not use templates from homologous PDB structures.  Since Rosetta uses fragments from existing PDB structures in order to guide the search in conjunction with energy functions, there is a semantic argument as to whether it is truly "ab initio" (although the same could be said for any statistically derived energy function).  Long story short: call it what you want, but be prepared for a debate!
</UL>
</P>


<br>
<a name="cm_semantics">
<H1 class="title">Q. What is the difference between Comparative and Homology Modeling? &nbsp; <a href="#top"><img src="../images/top.gif" border=0></a> </H1>
<UL>
<P>
Comparative vs. Homology Modeling: same thing, namely model a significant fraction of your target using coordinates from a homologous parent PDB structure.
</UL>
</P>




<BR><BR><BR>
<BR><BR><BR>
<BR><BR><BR>
<BR><BR><BR>